### PR TITLE
chore: store conflictings finality sigs in storage

### PR DESCRIPTION
**Pull request by gusin13**
## Description

Fixes - https://github.com/babylonlabs-io/rollup-bsn-contracts/issues/44 (store conflicting sigs)
Fixes - https://github.com/babylonlabs-io/rollup-bsn-contracts/issues/74 (contract should return err if duplicate sigs received)

## Checklist

- [x] I have updated the [SPEC.md](https://github.com/babylonlabs-io/rollup-bsn-contracts/blob/main/SPEC.md) file if this change affects the specification
- [x] I have updated the schema by running `cargo gen-schema`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -39,36 +39,62 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 
 ### State and API breaking
```
Comment (gusin13) [inline CHANGELOG.md:41]: changelog file had >80chars per line, so I formatted

File: e2e/main_test.go [hunk 2]
```diff
@@ -205,6 +222,62 @@ func (s *FinalityContractTestSuite) Test4SubmitFinalitySignature() {
...
+	blockToVote := sharedTestData.BlockToVote
+	appHash := sharedTestData.AppHash
+	idx := sharedTestData.Idx
+
+	signingCtx := signingcontext.FpFinVoteContextV0(s.ctx.ChainID(), s.contractAddr.String())
+	msgToSign := append([]byte(signingCtx), sdk.Uint64ToBigEndian(startHeight)...)
+	msgToSign = append(msgToSign, appHash...)
+
+	// Generate the same EOTS signature as Test4
+	sig, err := eots.Sign(fpSK, randListInfo.SRList[idx], msgToSign)
```
[the remark this answers is not in the document]
Comment (gusin13) [reply e2e/main_test.go:262]: it is already initialized at L137, maybe false positive from copilot

File: contracts/finality/src/state/finality.rs [hunk 3]
```diff
@@ -57,6 +73,7 @@ pub fn insert_signatories_by_block_hash(
 /// FinalitySigInfo is a struct that contains the finality signature and
 /// block hash for a given block height and fp
 #[cw_serde]
+#[derive(Hash, Eq)]
```
[the remark this answers is not in the document]
Comment (gusin13) [reply contracts/finality/src/state/finality.rs:76]: i dont think we need `partialEq`, `Clone` traits, seems like they are already provided by `cw_serde` 

if there was an issue related to traits then following would not have worked

no complain from cargo check
```
gusin@Gurjots-MacBook-Pro op-finality-gadget % cargo check
    Checking finality v0.1.0-dev (/Users/gusin/Github/labs/cursor-local-deployment-repos/op-finality-gadget/contracts/finality)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.38s
gusin@Gurjots-MacBook-Pro op-finality-gadget %
```

cargo build worked absolutely fine as well
```
gusin@Gurjots-MacBook-Pro op-finality-gadget % cargo build
   Compiling finality v0.1.0-dev (/Users/gusin/Github/labs/cursor-local-deployment-repos/op-finality-gadget/contracts/finality)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.79s
gusin@Gurjots-MacBook-Pro op-finality-gadget %
```

File: contracts/finality/src/exec/finality.rs [hunk 4]
```diff
@@ -29,15 +31,23 @@ pub fn handle_finality_signature(
 
     let fp_btc_pk = hex::decode(fp_btc_pk_hex)?;
 
-    // Load any type of existing finality signature by the finality provider at the same height
-    let existing_finality_sig = get_finality_signature(deps.storage, height, &fp_btc_pk)?;
+    // Load any existing finality signature by the finality provider at the same height
+    let existing_finality_sigs = list_finality_signatures(deps.storage, height, &fp_btc_pk)?;
```
Comment (SebastianElvis) [inline contracts/finality/src/exec/finality.rs:35]: Do you think we need a `get_finality_signature` function to get a finality signature over the given height/hash signed by the given BTC PK?

File: e2e/main_test.go [hunk 5]
```diff
@@ -205,6 +222,62 @@ func (s *FinalityContractTestSuite) Test4SubmitFinalitySignature() {
...
+	s.NoError(err)
+	votersResBz := s.QueryContract(s.contractAddr, string(queryVotersJson))
+	var votersRes BlockVotersResponse
+	err = json.Unmarshal(votersResBz, &votersRes)
+	s.NoError(err)
+	s.Equal(1, len(votersRes))
+	s.Equal(fpBTCPK.MarshalHex(), votersRes[0].FpBtcPkHex)
+}
+
+func (s *FinalityContractTestSuite) Test4_SubmitFinalitySigDuplicate() {
```
Comment (SebastianElvis) [inline e2e/main_test.go:243]: I don't think we need to test this in e2e because it is an edge case that does not involve Babylon layer

**Review by SebastianElvis — APPROVED** [review]

File: contracts/finality/src/exec/finality.rs (hunk 4 above)
Comment (gusin13) [reply contracts/finality/src/exec/finality.rs:35]: Ah you mean user would provide (pk,height,block_hash) combination? 

hm don't know where would this be used, seems like nice to have query but the current code don't need this. 

the votes storage we have is like this (pk,height) -> hashset(fp_info) 

where fp_info object is (sig,block_hash)

Comment (gusin13) [reply contracts/finality/src/exec/finality.rs:35]: contract doesn't have evidence store like Babylon, so the idea was to store all sigs in store if later offchain finality gadget wants to query and fetch the original vote or conflicting vote to make any decision

more context [here](https://babylonlabsworkspace.slack.com/archives/C07EDCYR77D/p1752280830707489?thread_ts=1752277341.256689&cid=C07EDCYR77D) and [here](https://github.com/babylonlabs-io/rollup-bsn-contracts/pull/41#discussion_r2181997857)

Comment (gitferry) [reply contracts/finality/src/exec/finality.rs:35]: Ah this is for rollup. I thought this is for cosmos integration. All good then.

**Review by gitferry — APPROVED** [review]
Great work! Will rollup fp need to adapt this change (handle duplicate error)? cc @Lazar955

**Comment by gusin13** [comment]
> Great work! Will rollup fp need to adapt this change (handle duplicate error)? cc @Lazar955

hm right, I think some changes needed here in rollup fp controller ([code ref](https://github.com/babylonlabs-io/finality-provider/blob/21b4abc0132742cda12b78f187f58433c7835a76/bsn/rollup-finality-provider/clientcontroller/consumer.go#L229)), the expected err strings there are aligned with babylon but these strings can be different in smart contract. 
Created a ticket to track https://github.com/babylonlabs-io/finality-provider/issues/521
